# Combinatoric Selections
https://projecteuler.net/problem=53

# Problem Statement
There are exactly ten ways of selecting three from five, 12345:

123, 124, 125, 134, 135, 145, 234, 235, 245, and 345

In combinatorics, we use the notation, 
5C3 = 10.

In general, nCr = n!/(r! * (n-r)!)
 
, where , r<= n , n! = n * (n-1) * (n-2) *...* 1 and 0! = 1.

It is not until n=23, that a value exceeds one-million: 23C10 = 1144066
.

How many, not necessarily distinct, values of nCr
 for 1 <= n <= 100, are greater than one-million?

# Planning
Write a function that calculates factorial

Write a function that calculates numbers of combinations

Write a function that tests values of n from 23 to 100 that have combination numbers greater than 1,000,000

In [1]:
def factorial(n):
    
    product = 1
    
    if n == 0:
        return product
    else:
        for x in range(1, n+1):
            product *= x
            
        return product

In [2]:
## testing
factorial(6)

720

In [7]:
def combinations(n, r):
    
    if r > n or r < 0:
        raise Exception("Improper r value")
    else:
        return int(factorial(n) / factorial(r) / factorial(n-r))

In [8]:
## testing
combinations(23, 10)

1144066

In [9]:
## testing flow of combination counting
n = 23

num_r_successful = 0 
over_1000000 = False

for r in range(1, int(n/2)+1):
    if combinations(n, r) > 1000000:
        over_1000000 = True
        break
        
if over_1000000:
    if n%2== 0: # if n is even
        num_r_successful = 2*( int(n/2) - r ) + 1
    else:
        num_r_successful = 2*( int(n/2) + 1 - r)
else:
    num_r_successful = 0
    
    
print(num_r_successful)
## includes 23C10, 23C11, 23C12, 23C13

4


In [10]:
## solution
num_r_successful_total = 0

for n in range(23, 100+1):
    num_r_successful = 0 
    over_1000000 = False

    for r in range(1, int(n/2)+1):
        if combinations(n, r) > 1000000:
            over_1000000 = True
            break

    if over_1000000:
        if n%2== 0: # if n is even
            num_r_successful = 2*( int(n/2) - r ) + 1
        else:
            num_r_successful = 2*( int(n/2) + 1 - r)
    else:
        num_r_successful = 0
        
    num_r_successful_total += num_r_successful
    
print(num_r_successful_total)

4075


# Solution
There are 4075 values of nCr greater than 1000000, not necessarily distinct, such that 1<=n<=100, and 0<=r<=n